src/core/net/ip6.cpp

```diff
@@ -603,27 +622,240 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+
+    while (payloadLeft != 0)
+    {
+        if (payloadLeft < maxPayloadFragment)
+        {
+            fragmentHeader.ClearMoreFlag();
+            payloadFragment = payloadLeft;
+            payloadLeft     = 0;
+            otLogDebgIp6("Last Fragment");
+        }
+        else
+        {
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        VerifyOrExit(fragment->Write(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetOffset(aMessage.GetOffset()));
+        VerifyOrExit(fragment->Write(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) ==
+                         sizeof(fragmentHeader),
+                     error = OT_ERROR_NO_BUFS);
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset() + (offset << 3),
+                                     aMessage.GetOffset() + sizeof(fragmentHeader), payloadFragment,
+                                     *fragment) == (int)payloadFragment,
+                     error = OT_ERROR_NO_BUFS);
+        EnqueueDatagram(*fragment);
+
+        fragmentCnt++;
+        fragment = NULL;
+        otLogInfoIp6("Fragment %d with %d bytes sent ", fragmentCnt, payloadFragment);
+    }
+exit:
+    aMessage.Free();
+
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        otLogWarnIp6("No buffer for Ip6 fragmentation");
+    }
+
+    if (error != OT_ERROR_NONE && fragment != NULL)
+    {
+        fragment->Free();
+    }
+#else
+    OT_UNUSED_VARIABLE(aMessage);
+    OT_UNUSED_VARIABLE(aIpProto);
+#endif // OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    return error;
+}
+
+otError Ip6::HandleFragment(Message &aMessage, Header &aHeader, MessageInfo &aMessageInfo)
 {
     otError        error = OT_ERROR_NONE;
     FragmentHeader fragmentHeader;
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    uint16_t offset          = 0;
+    uint16_t payloadFragment = 0;
+#endif // OPENTHREAD_ENABLE_IP6_FRAGMENTATION
 
     VerifyOrExit(aMessage.Read(aMessage.GetOffset(), sizeof(fragmentHeader), &fragmentHeader) == sizeof(fragmentHeader),
                  error = OT_ERROR_PARSE);
 
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    if (fragmentHeader.GetOffset() == 0 && fragmentHeader.IsMoreFlagSet() == false)
+    {
+        error = aMessage.MoveOffset(sizeof(fragmentHeader));
+    }
+    else if (mFragmentBuffer.mFragmentIdentification == 0 ||
+             (fragmentHeader.GetIdentification() == mFragmentBuffer.mFragmentIdentification &&
+              fragmentHeader.GetNextHeader() == mFragmentBuffer.mFragmentNextHeader))
```

Comment:
This implementation seems to only allow reassembly of one IPv6 datagram at a time. Would it be worthwhile to support multiple simultaneous reassembly buffers? You could do this by appending the reassembly state within the message buffer itself and maintaining a queue of reassembly messages outstanding.

Reply:
Yes, that's right. This implementation only allows the reassembly of one datagram at a time. Of course, it can be useful to support simultaneous reassembly. 
I don't know if I got that right. Could you clarify your proposal a little more, please?
An other simple solution would be to create the fragment buffer as an array whose size can be defined in the core config file. However, this may not be optimal due to the statically allocated memory. Thoughts?

Reply:
Rather than having a single message pointer, you could declare a `MessageQueue` that maintains a list of messages currently being reassembled.

I think we could avoid having to maintain most of `FragmentBuffer` by simply reading the information back from the message buffer. The exceptions that stand out are the Identification and Timeout values.

Fortunately, [Message::mTimeout](https://github.com/openthread/openthread/blob/af46ffa53b7d8cf17259dbb15f2b42b0d6573841/src/core/common/message.hpp#L97) already exists and something this can leverage.

For the Identification tag, [Message::mDatagramTag](https://github.com/openthread/openthread/blob/af46ffa53b7d8cf17259dbb15f2b42b0d6573841/src/core/common/message.hpp#L93) also exist, but needs to be expanded to 32 bits.

See [MeshForwarder::HandleFragment()](https://github.com/openthread/openthread/blob/af46ffa53b7d8cf17259dbb15f2b42b0d6573841/src/core/thread/mesh_forwarder.cpp#L1248) for an example of how 6lowpan reassembly works.

Thoughts?

Reply:
Yes, good solution! Thank you for the explanation.

Commit https://github.com/openthread/openthread/pull/3948/commits/ccecde1d786f5f16a006357cf9934a739561bd07 should now allow the multi-simultaneous assembly of IPv6 packets.